### PR TITLE
[Core] Remove unnecessary copy of multi modal input embeddings

### PR DESCRIPTION
**Pull request by lgeiger**
This improves throughput by about ~1% on some smaller models

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -1176,9 +1176,9 @@ def execute_model(
                     input_ids, mm_embeds)
             else:
                 inputs_embeds = self.model.get_input_embeddings(input_ids)
-            # TODO(woosuk): Avoid the copy. Optimize.
-            self.inputs_embeds[:num_scheduled_tokens].copy_(inputs_embeds)
-            inputs_embeds = self.inputs_embeds[:num_input_tokens]
+            if num_scheduled_tokens != num_input_tokens:
```
Comment (houseroad) [inline vllm/v1/worker/gpu_model_runner.py:1182]: I didn't get this part, why we can skip the copy if num_scheduled_tokens == num_input_tokens. Could you elaborate a bit?

Comment (lgeiger) [reply vllm/v1/worker/gpu_model_runner.py:1182]: You're right, I should've elaborated on this in the PR description. Here's my understanding of the code:

- `num_input_tokens` is computed here:
https://github.com/vllm-project/vllm/blob/432ec9926ebd3ce826f0d49df0a2a5ae3cc81ec0/vllm/v1/worker/gpu_model_runner.py#L1145-L1164
- So this code simply pads `input_embeds` to `num_input_tokens` if necessary but does it via `self.inputs_embeds` instead of `torch.nn.functional.pad` probably because of performance reasons.
- This means if `num_scheduled_tokens == num_input_tokens` and `self.inputs_embeds` is only used needed [dummy input](https://github.com/vllm-project/vllm/blob/432ec9926ebd3ce826f0d49df0a2a5ae3cc81ec0/vllm/v1/worker/gpu_model_runner.py#L1728) we don't actually need to do any padding and therefor can get rid of this copy.

Comment (njhill) [reply vllm/v1/worker/gpu_model_runner.py:1182]: Thanks @lgeiger, it would be great if you could include some comments with this explanation.

Comment (DarkLight1337) [reply vllm/v1/worker/gpu_model_runner.py:1182]: cc @WoosukKwon since you made the original comment

Comment (lgeiger) [reply vllm/v1/worker/gpu_model_runner.py:1182]: @njhill good point, added some explanation in 336c385ac143a8ebc7f37d4734fd0397e761c05b

Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:1182]: IIUC, we have to use the persistent buffer `self.inputs_embeds` (same as `self.input_ids` and `self.positions`) because of CUDA graphs. These buffers are used for CUDA graph capture, and has to be reused.

**Review by WoosukKwon — CHANGES_REQUESTED** [review]
IIUC, we have to use the persistent buffer `self.inputs_embeds` (same as `self.input_ids` and `self.positions`) because of CUDA graphs. These buffers are used for CUDA graph capture, and has to be reused.

File: vllm/v1/worker/gpu_model_runner.py (hunk 1 above)
Comment (lgeiger) [reply vllm/v1/worker/gpu_model_runner.py:1182]: Oh interesting, I wasn't aware. I didn't see any failures in my tests, sorry about that.
For the future, how could I reproduce such issues related to cuda graph capture?
